File: #0714.best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {

        int size = prices.size();
        int cash = 0;
        int price = -prices[0];
        for (int i = 1; i < size; i++) {
            cash = std::max(cash, price + prices[i] - fee);
            price = std::max(price, cash - prices[i]);
        }
        return cash;
    }
};

File: #0714.best-time-to-buy-and-sell-stock-with-transaction-fee.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function(prices, fee) {

    let result = 0;
    let buy =-prices[0] - fee;
    for (let i = 1; i < prices.length; i++) {
        if (prices[i] + buy > 0) {
            result += prices[i] + buy;
            buy =-prices[i];
        } else if (buy < -prices[i] - fee) {
            buy =- prices[i] - fee;
        }
    }
    return result;
};
